File: web/lib/score-ranges.ts
/**
 * Tetti per dimensione di uno score, lato web.
 *
 * L'autorità è `_validate_score_range` in `shared/skills/db_insert.py`: quello
 * è il solo writer che rifiuta un valore fuori scala. Il gemello JS per la CLI
 * è `cli/src/lib/score-ranges.js`; `tests/js/tasks/score-ranges.test.ts`
 * rilegge i tetti dal sorgente Python e fallisce se i righelli divergono.
 *
 * Nota sul totale: la somma dei tetti fa 110, non 100, mentre `--total` è
 * validato 0..100. La contraddizione è nota e non si chiude qui.
 */
export const SCORE_COMPONENT_LIMITS = {
  stack_match: 40,
  remote_fit: 25,
  salary_fit: 20,
  experience_fit: 10,
  strategic_fit: 15,
} as const;

export type ScoreComponent = keyof typeof SCORE_COMPONENT_LIMITS;

export interface OutOfRangeHit {
  column: ScoreComponent;
  value: number;
  max: number;
}

export interface OutOfRangeSummary {
  rows: number;
  byColumn: Partial<Record<ScoreComponent, number>>;
  worst: OutOfRangeHit | null;
}

/**
 * Elenca le dimensioni di una riga score che sfondano il proprio massimo.
 *
 * Non normalizza e non scarta nulla: serve a *contare*, non a correggere. I
 * punteggi già scritti appartengono a utenti reali, e un clamp silenzioso al
 * confine della sincronizzazione nasconderebbe il fenomeno invece di chiuderlo.
 */
export function outOfRangeComponents(
  row: Record<string, unknown> | null | undefined,
): OutOfRangeHit[] {
  if (!row || typeof row !== "object") return [];
  const hits: OutOfRangeHit[] = [];
  for (const [column, max] of Object.entries(SCORE_COMPONENT_LIMITS) as [
    ScoreComponent,
    number,
  ][]) {
    const value = row[column];
    if (typeof value !== "number" || !Number.isFinite(value)) continue;
    if (value > max || value < 0) hits.push({ column, value, max });
  }
  return hits;
}

/** Aggrega il conteggio fuori scala su un insieme di righe score. */
export function summarizeOutOfRange(
  rows: Iterable<Record<string, unknown>> | null | undefined,
): OutOfRangeSummary {
  const byColumn: Partial<Record<ScoreComponent, number>> = {};
  let affected = 0;
  let worst: OutOfRangeHit | null = null;
  for (const row of rows ?? []) {
    const hits = outOfRangeComponents(row);
    if (hits.length === 0) continue;
    affected++;
    for (const hit of hits) {
      byColumn[hit.column] = (byColumn[hit.column] ?? 0) + 1;
      // "Peggiore" = quanto sfonda in rapporto al proprio tetto, non il valore
      // assoluto: 18/15 e 45/40 sfondano di 3 punti ma non allo stesso modo.
      if (worst === null || hit.value / hit.max > worst.value / worst.max) {
        worst = hit;
      }
    }
  }
  return { rows: affected, byColumn, worst };
}

File: shared/skills/score_ranges.py
#!/usr/bin/env python3
"""Tetti per dimensione di uno score — la sola fonte, lato Python.

Prima questi cinque numeri stavano scritti a mano in `db_insert.py`, nei sette
prompt dello Scorer, in `tests/test_pipeline_agents.py`, nella pagina posizione
del web e nel pannello del gioco. Divergevano: l'esempio copiabile del prompt
ha insegnato `--experience-fit 20` contro un tetto di 10 per mesi, e nessuno
se n'è accorto perché ogni copia era corretta rispetto a sé stessa.

I gemelli JS sono `cli/src/lib/score-ranges.js` e `web/lib/score-ranges.ts`;
`tests/js/tasks/score-ranges.test.ts` rilegge da qui e fallisce se divergono.

⚠️ Contraddizione nota e NON risolta qui: la somma dei tetti fa 110, mentre
`--total` è validato 0..100. Le due regole «il totale è la somma delle
dimensioni» e «il totale sta in 0..100» non possono valere insieme in cima
alla scala. Chiuderla vuol dire ripesare le dimensioni, che è una decisione di
prodotto, non un riordino di questo file.
"""

from __future__ import annotations

# Ordine: quello in cui `db_insert.py score` li valida, per leggere il diff
# accanto al codice che li applica.
COMPONENT_LIMITS: dict[str, int] = {
    "stack_match": 40,
    "remote_fit": 25,
    "salary_fit": 20,
    "experience_fit": 10,
    "strategic_fit": 15,
}

#: Tetto del punteggio complessivo. Vedi la contraddizione nel docstring.
TOTAL_LIMIT = 100

#: Nome leggibile di ogni dimensione, per il testo di `--help` della CLI.
SCORE_COMPONENT_LABELS: dict[str, str] = {
    "stack_match": "Stack",
    "remote_fit": "Remote/location",
    "salary_fit": "Salary",
    "experience_fit": "Seniority",
    "strategic_fit": "Strategic",
}


def out_of_range(values: dict[str, int | None]) -> list[tuple[str, int, int]]:
    """Elenca le dimensioni fuori dal proprio intervallo `[0, tetto]`.

    Conta, non corregge: i punteggi già scritti appartengono a utenti reali e
    un clamp silenzioso nasconderebbe il fenomeno invece di chiuderlo.

    :returns: lista di ``(colonna, valore, tetto)``, vuota se tutto è in scala.
    """
    found = []
    for column, maximum in COMPONENT_LIMITS.items():
        value = values.get(column)
        if value is None or isinstance(value, bool):
            continue
        if not isinstance(value, int):
            continue
        if value < 0 or value > maximum:
            found.append((column, value, maximum))
    return found
